walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,99 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // if the request meeting is longer than the duration of one day
+    // return an empty list
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // if there is no event on the calendar currently
+    // return whole day as possible time slot
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    // if no mandatory attendees or optional attendees, return whole day
+    if (mandatoryAttendees.size() == 0 && optionalAttendees.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // get unavailable time of mandatory attendees
+    List<TimeRange> unavailableTime = getUnavailableTime(events, mandatoryAttendees);
+    // merge available intervals with overlaps
+    List<TimeRange> mergedUnavailableTime = mergeUnavailableTime(unavailableTime);
+    // get available time slot base on merged unavailable time
+    List<TimeRange> res = getAvailableTime(mergedUnavailableTime, request.getDuration());
+    return res;
+  }
+
+  private List<TimeRange> getUnavailableTime(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> res = new ArrayList<>();
```

Comment:
Nit: change `res` to be something like `unavailableTimes` so it's more descriptive.